### ISSUE TITLE
PGIP XML support

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#2705]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 2705
Date: 2012-02-10 18:17:02 +0100
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;
To:   
CC:   @_cpitclaudel, @_ejgallego, @_gares, @_JasonGross, @_pirbo, @_Matafou, @_Matafou, @_pstecker

Last updated: 2017-03-30 00:33:20 +0200

## Bugzilla Comment ID: 4801
Date: 2012-02-10 18:17:02 +0100
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;

To the best of my knowledge (I may be wrong), coqide uses a proprietary, OCaml marshalling based interface to communicate with Coq, while the current proofgeneral interface manually parses the output of coqtop -emacs, which is only semi-machine friendly.

It would be great if Coq had support for PGIP's XML marshalling interface. It would assist in creating alternate interfaces to Coq, and allow more experimentation.

(Let me know if this is a duplicate!)

Cheers,
Edward

## Bugzilla Comment ID: 4802
Date: 2012-02-10 19:20:58 +0100
From: @_pirbo

Why do you say that OCaml marshalling is propriatory ? 
Anyway, No coqide uses an adhoc xml to talk with coqtop and yes we want PG to stop using fragile string style communication.

I think none of us were aware that PGIP's XML exists and we should see the cost of implementing it. I suspect that it is a lot.
Futhermore I do not see easily in the link a formal exhaustive description of the language (I should go in the CVS ?). Can you give us one ?
This way, Pierre Marie who implements the xml could see if it can be done.

Pierre Courtieu might also have clues...

But this is where we want to go yes !

## Bugzilla Comment ID: 4803
Date: 2012-02-10 19:27:09 +0100
From: @_pirbo

Sorry page 6 of http://proofgeneral.inf.ed.ac.uk/Kit/docs/commentary.pdf is the begenning of our playground !!!!
We should play this game ...

## Bugzilla Comment ID: 4804
Date: 2012-02-10 19:31:06 +0100
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;

Sorry, when I said "proprietary", I meant "Coq-specific" language.

I think implementing PGIP would be nontrivial, though fortunately it's already been done with Isabelle, which is a good sign.

I thought the draft commentary was useful, but it may be out of date: http://proofgeneral.inf.ed.ac.uk/Kit/docs/commentary.pdf  ; and I think this is the formal exhaustive version: http://proofgeneral.inf.ed.ac.uk/Kit/dtd/pgip.rnc (view-source so that it gets formatted properly!)

We may also want to let David Aspinall know that Coq is attempting to adopt PGIP.

## Bugzilla Comment ID: 4921
Date: 2012-04-06 03:43:35 +0200
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;

Hello,

Was there any progress on this? If not, I think I might want to attempt taking a stab at implementing a subset of PGIP that I want for my application.

Cheers,
Edward

## Bugzilla Comment ID: 4922
Date: 2012-04-06 09:34:51 +0200
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;

Here is a branch for this functionality: https://github.com/ezyang/coq

Most of the commits are cleaning up code to properly handle XML-ized output. It's not very elegant but it works without needing to transform all of the codebase.

## Bugzilla Comment ID: 4927
Date: 2012-04-07 05:56:14 +0200
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;

I think these three commits are far along enough for code review:

commit d1c3193beb1d9ac8370280748116207246dafccf
Author: Edward Z. Yang <ezyang@mit.edu>
Date:   Fri Apr 6 23:26:56 2012 -0400

    Partial PGIP support.
    
    Signed-off-by: Edward Z. Yang <ezyang@mit.edu>

commit 5444cc0ff0c2ab6fcf6729c526caa0a595384d6a
Author: Edward Z. Yang <ezyang@mit.edu>
Date:   Fri Apr 6 01:28:59 2012 -0400

    Initial support for printing XML, and some new flags.
    
    Signed-off-by: Edward Z. Yang <ezyang@mit.edu>

commit 4cfd5a0569f23f782d0bedad023924e715c9b71b
Author: Edward Z. Yang <ezyang@mit.edu>
Date:   Fri Apr 6 00:42:33 2012 -0400

    Remove and annotate suspicious print calls. Did not check plugins.
    
    Signed-off-by: Edward Z. Yang <ezyang@mit.edu>

## Bugzilla Comment ID: 4928
Date: 2012-04-07 06:06:59 +0200
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;

Possible improvements to the current patches:

- Figure out whether or not msg/msgnl should automatically wrap input to it with tags, or if some of the code not in toplevel/printer should be refactored not to call this directly (as is the case with library/declare.ml)

- Maybe stop using physical equality on Print_emacs/co flags (what's the right alternative?)

- More comprehensive sweep to get all of the PGIP output; I've definitely only got a subset.  In particular, what code is responsible for outputting all of the tactics when you type 'Qed'?

- PGIP helper functions for setting up common tags, instead of calling "tag" manually. Maybe a tag0 helper function for when there are no attributes.

- Reorganize some of the changes; split up second patch into adding XML support, and adding the flags, and move all semantics changes to last patch.

- Bikeshed and document proprietary attributes (that aren't defined by PGIP)

However, I think these patches are good enough for my personal use, so I probably am not going to develop them further unless you tell me these should get fixed if I want the patches put into Coq.

## Bugzilla Comment ID: 4948
Date: 2012-04-12 23:22:43 +0200
From: @_pirbo

Hi,
Your first patch is applied in trunk.
I thing you are right to push us to move on and not wait for us. Moreover, I've learned that someone try to make a eclipse interface for Coq, he must take advantage of this work and advocate in favour to do it fast.

Anyway, I'm more and more confused about this issue and Pierre Courtieu does not answer !

Since my last message, I've learned the arguments against :
 - that isabelle refuses PGIP as a communication language to make its parallel interface. Coq is part of a common project with Isabelle whose goal is exactly doing things in parallel. Consequently, PGIP might not be the right candidate.
 - At the beginning of a top-down implementation of PGIP, 
 * start/stop the prover should be OK
 * interrupt by xml message instead of unix signal is a huge amount of low level  (windows nightmare) work
 * normal respond / error respond should be easy
 * we need to open and close proofs. But Coq needs to distinguish between Lemma and Fixpoint one the one hand and Qed and Defined on the other hand. There is only one way of doing these actions in PGIP and it is absolutely blocker.

Independently, I think easier to start from coqtop -ideslave (coqide mode) than from coqtop -emacs to implement the protocol but it may be incompatible with your work ...

Conclusion, No progress but I do try to tackle the issue :-)

## Bugzilla Comment ID: 4950
Date: 2012-04-12 23:40:17 +0200
From: Edward Z. Yang &lt;<ezyang@mit.edu>&gt;

Yes, I had a conversation with Makarius on the proof-general list about the unsuitability of PGIP as a protocol going forward. It seems that PGIP may not be the one true way, and maybe it will have to wait until a sufficiently motivated undergrad decides to design a new protocol. Alas, there seem to be a dearth of open research questions in such an endeavour.

I have not thought very carefully about how to handle interrupts using the in-band protocol.  The lack of threading in OCaml seems to be one major difficulty in handling this case sensibly; if we had threads we could just put the stdin-handler in its own thread and maintain responsiveness at all times.

It should not surprise you that my two patches do not deal at all with PGIP's interaction protocol; they only involve annotating the output of Coq with extra useful information. I think that such changes have quite a good effort to utility ratio: you add a few extra calls to the 'tag' combinator, and when you turn on the flag you get lots of extra info that makes parsing a bit easier.

## Bugzilla Comment ID: 5190
Date: 2012-06-20 11:59:12 +0200
From: @_Matafou

Hi, 

sorry for being so unreactive. I do not read Coq bugs very often.

Indeed PGIP is rather outdated now. I know of no real successful interface using PGIP. Isabelle almost made it and the switched to something else.

It is still a good idea to switch to an xml communication format between coq and its interfaces. I would encourage any effort in this direction. Coqide is doing this and we could try to use the same format with pg. For the moment anyway the communication between proofgeneral and coq is very simple: it is the standard text interface with slight modifications: mainly some extra information put in the prompt (with a special ugly syntax for that).

Changing this interaction to xml would be easy and I will probably do it when I have time for it (which may be never). Changing to a complete xml communication protocol (i.e. terms and goals description would be sent in xml) is more work and I will not do it before having a real stable protocol for that.

This is where the real problem comes: to come up with a good *generic* protocol meeting the requirements of today (mainly parallel/asynchronous communication). Let us say this could but PGIP 2...

In project paralITP Makarius Wenzel is sharing with us his experience of Isabelle/Scala interface, which is great. However the protocol he is using is not clear to me currently. It seems to be an adhoc protocol specifically designed to have asynchonous computations *and* reactivity (parallel computations seems to be difficult). Making this protocol more generic is one of the challenges of paralITP but it is not clear yet if we will succeed.

Hope this helps,
P.

## Bugzilla Comment ID: 13074
Date: 2016-07-11 12:22:52 +0200
From: @_JasonGross

Should this be closed?  CoqIDE uses an XML protocol, now, and there are other protocols in the works.

## Bugzilla Comment ID: 14381
Date: 2016-10-17 15:30:09 +0200
From: @_ejgallego

I'd vote for closing indeed.

## Bugzilla Comment ID: 15689
Date: 2017-03-30 00:33:20 +0200
From: @_ejgallego

Using my newly minted powers I am closing this as WONTFIX. Users interested in interaction with Coq can use the XML protocol or SerAPI.
